### PR TITLE
Solve connection handling

### LINKED ISSUE #101: Solve connection handling
There are a few different scenarios to support with v1 backcompat:

1. Someone installs device-scanner pre-IML@4.0.4. Their `chroma-agent` will write `{"ACTION": "info"}` to the socket and then half-close it immediately.
2. Someone installs device-scanner at or past IML@4.0.4. They use `scanner-proxy` to maintain a persistent connection, that first writes `"Info"\n` at startup.

In addition, the listeners for Udev and ZED will write changes to the `device-scanner`, they use a connection that can end immediately after writing (they don't even wait for a connection to `device-scanner`, they just write their data to the unix domain socket and exit).

The current way of modeling this is with `Persistent` and `Ephemeral` connections. a `V2` `Info` command is `Persistent`, everything else is `Ephemeral`. However, this abstraction doesn't solve the problem, as the `V1` command gets lumped in with all the listener commands. Because of this, it's not possible to know what type of data each connection expects, and the result is connections will get data fed to them based on the type of the last connected socket.

To solve this, we need to move the query side closer to the connection.

The rules should be:

1. a V1 info request gets a V1 info response
2. a V2 info request gets a V2 info response
3. everything else gets no response as they are just feeding new data.

### PR DESCRIPTION
**Pull request by jgrund**
Fixes #101.

There are a few different scenarios to support with v1 backcompat:

Someone installs device-scanner pre-IML@4.0.4. Their chroma-agent will write `{"ACTION": "info"}` to the socket and then half-close it immediately.
Someone installs device-scanner at or past IML@4.0.4. They use scanner-proxy to maintain a persistent connection, that first writes `"Info"\n` at startup.

In addition, the listeners for Udev and ZED will write changes to the device-scanner, they use a connection that can end immediately after writing (they don't even wait for a connection to device-scanner, they just write their data to the unix domain socket and exit).

The current way of modeling this is with Persistent and Ephemeral connections. a V2 Info command is Persistent, everything else is Ephemeral. However, this abstraction doesn't solve the problem, as the V1 command gets lumped in with all the listener commands. Because of this, it's not possible to know what type of data each connection expects, and the result is connections will get data fed to them based on the type of the last connected socket.

To solve this, we need to move the query side closer to the connection.

The rules should be:

- a V1 info request gets a V1 info response
- a V2 info request gets a V2 info response
- everything else gets no response as they are just feeding new data.

Signed-off-by: Joe Grund <joe.grund@intel.com>

### CODE REVIEW

File: IML.DeviceScannerDaemon/src/Connections.fs [hunk 1]
```diff
@@ -6,42 +6,60 @@ module IML.DeviceScannerDaemon.Connections
...
+  | V2 of Net.Socket
+  | ReadOnly of Net.Socket
 
 let mutable conns:Connection list = []
 
-let removeConn (c) () =
+let addConn c =
+  conns <- c :: conns
+
+let removeConn (c) =
```
Comment (tanabarr) [inline IML.DeviceScannerDaemon/src/Connections.fs]: why the brackets around c?

File: IML.DeviceScannerDaemon/test/__snapshots__/ConnectionsTest.fs.snap [hunk 2]
```diff
@@ -0,0 +1,69 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`writing a connection 1`] = `
+Object {
+  "data": Array [
```
Comment (tanabarr) [inline IML.DeviceScannerDaemon/test/__snapshots__/ConnectionsTest.fs.snap]: why not match against string version of buffer to make the snapshot more human readable

File: IML.DeviceScannerDaemon/src/Connections.fs (hunk 1 above)
Comment (jgrund) [reply IML.DeviceScannerDaemon/src/Connections.fs]: I'll remove.

File: IML.DeviceScannerDaemon/test/__snapshots__/ConnectionsTest.fs.snap (hunk 2 above)
Comment (jgrund) [reply IML.DeviceScannerDaemon/test/__snapshots__/ConnectionsTest.fs.snap]: Sure.

Comment (jgrund) [reply IML.DeviceScannerDaemon/test/__snapshots__/ConnectionsTest.fs.snap]: Ok, added a serializer for buffers.

File: IML.DeviceScannerDaemon/src/Connections.fs (hunk 1 above)
Comment (jgrund) [reply IML.DeviceScannerDaemon/src/Connections.fs]: Fixed.

**Pull request merged**